src/transformStateWithClones.js

```
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+      case 'clear':
+        currentState = {};
+        break;
+      default:
+        return 'error data';
```

This default case handles unknown action types by returning the string `'error data'`. This breaks the function contract (the function should return an array of state objects) and also exits early without pushing the state for the current action. It is better to explicitly throw an Error (so failures are clear) or to document and implement a chosen strategy (e.g., skip unknown actions but still push the current state). For example:

```javascript
default:
  throw new Error(`Unknown action type: ${action.type}`);
```

This aligns with the checklist guidance about error handling and ensures the caller receives a clear failure instead of a wrong return type.